src/storage/hummock_sdk/src/time_travel.rs

```diff
@@ -0,0 +1,348 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use risingwave_common::catalog::TableId;
+use risingwave_pb::hummock::group_delta::DeltaType;
+use risingwave_pb::hummock::hummock_version::PbLevels;
+use risingwave_pb::hummock::hummock_version_delta::{ChangeLogDelta, PbGroupDeltas};
+use risingwave_pb::hummock::{
+    EpochNewChangeLog, PbGroupDelta, PbHummockVersion, PbHummockVersionDelta, PbIntraLevelDelta,
+    PbLevel, PbOverlappingLevel, PbSstableInfo, SstableInfo, StateTableInfoDelta,
+};
+
+use crate::change_log::TableChangeLog;
+use crate::table_watermark::TableWatermarks;
+use crate::version::{HummockVersion, HummockVersionDelta, HummockVersionStateTableInfo};
+use crate::{CompactionGroupId, HummockSstableId};
+
+/// [`IncompleteHummockVersion`] is incomplete because `SSTableInfo` only has the `sst_id` set in the following fields:
+/// - `PbLevels`
+/// - `TableChangeLog`
+#[derive(Debug, Clone, PartialEq)]
+pub struct IncompleteHummockVersion {
+    pub id: u64,
+    pub levels: HashMap<CompactionGroupId, PbLevels>,
+    pub max_committed_epoch: u64,
+    safe_epoch: u64,
+    pub table_watermarks: HashMap<TableId, Arc<TableWatermarks>>,
+    pub table_change_log: HashMap<TableId, TableChangeLog>,
+    pub state_table_info: HummockVersionStateTableInfo,
+}
+
+/// Clone from an `SstableInfo`, but only set the `sst_id` for the target, leaving other fields as default.
+/// The goal is to reduce the size of pb object generated afterward.
+fn stripped_sstable_info(origin: &SstableInfo) -> SstableInfo {
+    SstableInfo {
```

Comment:
This is a note instead of a comment. At the first glance, I think we can just have a new struct with sstable_info to be an `Option` instead of still writing 12 empty fields, which may occupy the space unneccessarily. I then realized that this is because to do so we need a new struct for each nested pb messages in HummockVersion including `Levels`, `Level`, `OverlappngLevel` and ... Eventually I think we should get rid of using PBs directly in meta and switch to our own struct (probably with Arc on sstable info as well to save memory) but yeah we don't need to do that in this PR.